### PR TITLE
Reland "Add torch.utils.device_mode"

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #91892
* __->__ #91796

Original PR https://github.com/pytorch/pytorch/pull/91525

Signed-off-by: Edward Z. Yang <ezyangfb.com>

### CODE REVIEW

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42381089).

**Review by albanD — COMMENTED** [review]
Are you planning internal-only changes for this?

**Comment by ezyang** [comment]
No, let me mark the difference

File: torch/utils/_contextlib.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+class _DecoratorContextManager:
+    """Allow a context manager to be used as a decorator"""
+
+    def __call__(self, orig_func: F) -> F:
+        if inspect.isclass(orig_func):
+            warnings.warn("Decorating classes is deprecated and will be disabled in "
+                          "future versions. You should only decorate functions or methods. "
+                          "To preserve the current behavior of class decoration, you can "
+                          "directly decorate the `__init__` method and nothing else.")
+            func = cast(F, lambda *args, **kwargs: orig_func(*args, **kwargs))
```
Comment (ezyang) [inline torch/utils/_contextlib.py:129]: Previously I overwrote func in the frame, now the closed over func has its own name

File: torch/_C/__init__.pyi.in [hunk 2]
```diff
@@ -45,6 +45,12 @@ class device:
     @overload
     def __init__(self, type: str, index: _int) -> None: ...
 
+    def __call__(self, func: T) -> T: ...
```
Comment (albanD) [inline torch/_C/__init__.pyi.in:48]: there is no call anymore.

**Review by albanD — APPROVED** [review]

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42381089).

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42381089).

File: test/test_autograd.py [hunk 3]
```diff
@@ -122,22 +122,36 @@ def test_grad_mode_class_decoration(self):
...
+                def __init__(self):
+                    assert not torch.is_grad_enabled()
+
+                def foo(self):
+                    # Not applied to methods
+                    assert torch.is_grad_enabled()
+
+            # Show that we can actually construct the class
+            foo = Foo()
+            foo.foo()
```
Comment (ezyang) [inline test/test_autograd.py:134]: Improved test here now catches the infinite loop bug from previous

**Comment by ezyang** [comment]
These tests don't fail locally for me with the PR.

**Pull request closed without merging**